src/operator/fusion/fused_op.cu

```diff
@@ -0,0 +1,396 @@
+#include "../operator_common.h"
+#include "../elemwise_op_common.h"
+#include "../../executor/exec_pass.h"
+#include "../../common/cuda_utils.h"
+
+namespace mxnet {
+
+namespace detail {
+
+inline std::string mshadowTypeToString(int type) {
+  switch (type) {
+    case mshadow::kFloat32:
+      return "float";
+    case mshadow::kFloat64:
+      return "double";
+    case mshadow::kFloat16:
+      return "half";
+    case mshadow::kUint8:
+      return "unsigned char";
+    case mshadow::kInt8:
+      return "char";
+    case mshadow::kInt32:
+      return "int";
+    case mshadow::kInt64:
+      return "long long";
+    default:
+      LOG(FATAL) << "Unknown type enum " << type;
+  }
+  return "";
+}
+
+}  // namespace detail
+
+void FusedOp::GenerateCode(const std::vector<OpReqType> &req) {
+  const auto& g = this->symbol_.indexed_graph();
+  std::string code = "";
+  int temp_name_counter = 0;
+  using NodeEntry = nnvm::IndexedGraph::NodeEntry;
+  std::map<std::pair<int, int>, std::string> variables;
+
+  std::vector<uint32_t> outputs(g.num_nodes());
+
+  for (size_t i = 0; i < g.num_nodes(); ++i) {
+    const auto& node = g[i];
+    if (node.source != nullptr) {
+      outputs[i] = node.source->num_outputs();
+    } else {
+      outputs[i] = 0;
+    }
+  }
+
+  for (size_t i = 0; i < g.num_nodes(); ++i) {
+    const auto& node = g[i];
+    const auto* source = node.source;
+    if (source != nullptr) {
+      std::string var_name = "temp" + std::to_string(temp_name_counter++);
+      if (source->is_variable()) {
+        code += "const auto " + var_name + " = load(" + source->attrs.name + ", i);\n";
+        CHECK_EQ(outputs[i], 1);
+        variables[{i, 0}] = var_name;
+      } else {
+        std::string op_name = source->op()->name;
+        if (detail::fused_op_binary_ops.find(op_name) != detail::fused_op_binary_ops.end()) {
+          std::string op = detail::fused_op_binary_ops.at(op_name);
+          const auto& arg1 = variables[{node.inputs[0].node_id, node.inputs[0].index}];
+          const auto& arg2 = variables[{node.inputs[1].node_id, node.inputs[1].index}];
+          code += "const auto " + var_name + " = " + op +
+                  "(" + arg1 + ", " + arg2 + ");\n";
+          CHECK_EQ(outputs[i], 1);
+          variables[{i, 0}] = var_name;
+          continue;
+        }
+
+        if (detail::fused_op_unary_ops.find(op_name) != detail::fused_op_unary_ops.end()) {
+          std::string op = detail::fused_op_unary_ops.at(op_name);
+          const auto& arg1 = variables[{node.inputs[0].node_id, node.inputs[0].index}];
+          code += "const auto " + var_name + " = " + op +
+                  "(" + arg1 + ");\n";
+          CHECK_EQ(outputs[i], 1);
+          variables[{i, 0}] = var_name;
+          continue;
+        }
+
+        if (detail::fused_op_special_ops.find(op_name) != detail::fused_op_special_ops.end()) {
+          const std::vector<std::string>& op_desc = detail::fused_op_special_ops.at(op_name);
```

Comment:
why not reuse the iterator that you have when doing find? this is accessing the map two times.

Reply:
Then I would need to have the iterator in the outer scope, which I think is less readable (as the name would need to be different for each one or another scope would need to be made). And this function is not used in the performance-critical section of the run.